File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/19-TopologicalSort/LeetCode/LeetCode-0207-2.cpp
class Solution {
#define BLACK 0
#define GRAY  1
#define WHITE 2
    
public:   
    ///--------------------------------------------------------------
    /// Do DFS & look for back edges (loops)
    ///
    bool dfs(int u, vector<vector<int>>& G, vector<int> &color){
        color[u] = GRAY;
        
        for (int v: G[u]){
            if (color[v] == GRAY) 
                return false;  // Backedge
            if (color[v] == WHITE){
                if (!dfs(v, G, color))
                    return false;
            } //end-if
        } // end-for

        color[u] = BLACK;
        
        return true;
    } //end-dfs
                
    ///-----------------------------------------------------------------
    /// Convert the prerequisites to a graph. 
    /// Then do a DFS & look for a backedge (llop) 
    ///
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // Create DAG from prerequisites
        vector<vector<int>> G(numCourses);
        
        for (int i=0; i<prerequisites.size(); i++)
            G[prerequisites[i][1]].push_back(prerequisites[i][0]);
        
        // Do DFS & look for cycles (back edges)        
        vector<int> color(numCourses, WHITE);
        
        for (int i=0; i<numCourses; i++)
            if (color[i] == WHITE)
                if (!dfs(i, G, color))
                    return false;
        
        return true;
    } // end-numCourses
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/19-TopologicalSort/LeetCode/LeetCode-0207-2.py
class Solution:
    BLACK = 0
    GRAY = 1
    WHITE = 2

    #--------------------------------------------------------------
    # Do DFS & look for back edges (loops)
    #
    def dfs(self, u, G, color):
        color[u] = Solution.GRAY
        
        for v in G[u]:
            if (color[v] == Solution.GRAY):
                return False         # Backedge
            if (color[v] == Solution.WHITE):
                if (not self.dfs(v, G, color)):
                    return False

        color[u] = Solution.BLACK        
        return True

    #-----------------------------------------------------------------
    # Convert the prerequisites to a graph. 
    # Then do a DFS & look for a backedge (llop) 
    #                
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # Create DAG from prerequisites
        G = []
        for i in range(numCourses): G.append([])
            
        for i in range(len(prerequisites)):
            G[prerequisites[i][1]].append(prerequisites[i][0])
                
        # Do DFS & look for cycles (back edges)        
        color = [Solution.WHITE]*numCourses
        print(color)
        
        for u in range(numCourses):
            if (color[u] == Solution.WHITE):
                if (not self.dfs(u, G, color)):
                    return False
        
        return True
        